### ISSUE TITLE
Refactor gevent.subprocess for Py2/Py3?

### ISSUE THREAD

**Issue by jamadden** [gevent/gevent#748]
Right now, gevent.subprocess has a large number of  `if PY3` statements, and the like. Together with the stdlib's `if windows` branches, this leads to a large number of combinations and makes the code somewhat hard to follow. Plus, it means that the generated documentation is only correct for the platform that generated it (we can't document both interfaces cleanly).

Therefore, is it worth splitting gevent.subprocess into a `_subprocess2.py` and `_subprocess3.py`, as was done for `gevent.socket` and `gevent.ssl`? The benefits would be a simpler, easier to read control flow and improved documentation. The downside would be a great deal of nearly-identical code, and there's some benefit to the current situation where you can examine one file to see *exactly* what the behaviour differences are.

**Comment by thedrow** [comment]
We can split it to `_subprocess2.py`, `_subprocess3.py` and keep the shared code in `subprocess.py` if that's possible.

**Comment by jamadden** [comment]
In #883, I added an implementation of `subprocess.run` to gevent and in the process made it available as an extension to Python <= 3.4 (previously it was new to Python 3.5).

In doing so, I believe I came to the conclusion that it's reasonable to offer extensions from newer python versions to older python versions *so long as there are no (or extremely limited) compatibility concerns*. For example, a `Popen` object can now be used as a context manager in Python 2; trying that before would have immediately failed with a `TypeError`, so any backwards compatibility concerns are extremely limited to nearly non-existant.

Using that, I suspect we can simplify the subprocess module substantially and eliminate many `if PY3` checks.
